### PR TITLE
buffer filter: Populate content-length header

### LINKED ISSUE #7844: buffer: Synthesize `content-length` header in case of request buffering
*Description*:
When envoy processes request with enabled buffering it reads the whole request into memory before sending headers to an upstream.
Is is possible to make Envoy populate `content-length` header in this case? It would help with some http/1.0 backends (which unfortunately still exist in the wild).

Another option would be to add http/1.0-compliant behavior as cluster option (based on the precedent of existing `accept_http10` option).

Without either of this options Envoy tries to send chunked-encoded data, which is observed as empty body by http/1.0 server.

*Relevant Links*:
* https://www.envoyproxy.io/docs/envoy/latest/configuration/http_filters/buffer_filter

### PR DESCRIPTION
**Pull request by euroelessar**
Description: Make buffer filter populate `content-length` header if it is not present in the request already. As buffer has to fetch the whole request in order to proceed, calculation of the header value involves mostly no extra cost.
Risk Level: MEDIUM
Testing: added unit and integration tests
Docs Changes: added section about new behavior around `content-length` header
Release Notes: updated
Fixes #7844

### CODE REVIEW

File: docs/root/intro/version_history.rst [hunk 1]
```diff
@@ -6,6 +6,7 @@ Version history
 * access log: added :ref:`buffering <envoy_api_field_config.accesslog.v2.CommonGrpcAccessLogConfig.buffer_size_bytes>` and :ref:`periodical flushing <envoy_api_field_config.accesslog.v2.CommonGrpcAccessLogConfig.buffer_flush_interval>` support to gRPC access logger. Defaults to 16KB buffer and flushing every 1 second.
 * admin: added ability to configure listener :ref:`socket options <envoy_api_field_config.bootstrap.v2.Admin.socket_options>`.
 * admin: added config dump support for Secret Discovery Service :ref:`SecretConfigDump <envoy_api_msg_admin.v2alpha.SecretsConfigDump>`.
+* buffer filter: the buffer filter populates content-length header if not present, behavior can be disabled using the runtime feature `envoy.reloadable_features.buffer_populate_content_length`.
```
Comment (mattklein123) [inline docs/root/intro/version_history.rst]: I would call this `envoy.reloadable_features.buffer_filter_populate_content_length`. Also, can you refer to this in the fitler docs also?

File: source/extensions/filters/http/buffer/buffer_filter.cc [hunk 2]
```diff
@@ -70,12 +71,20 @@ Http::FilterHeadersStatus BufferFilter::decodeHeaders(Http::HeaderMap&, bool end
...
+  request_headers_ = &headers;
 
   return Http::FilterHeadersStatus::StopIteration;
 }
 
-Http::FilterDataStatus BufferFilter::decodeData(Buffer::Instance&, bool end_stream) {
+Http::FilterDataStatus BufferFilter::decodeData(Buffer::Instance& data, bool end_stream) {
+  content_length_ += data.length();
   if (end_stream || settings_->disabled()) {
+    if (request_headers_ != nullptr && request_headers_->ContentLength() == nullptr) {
```
Comment (mattklein123) [inline source/extensions/filters/http/buffer/buffer_filter.cc]: I see that request_headers_ being non-null is guarded by disabled so the if statement above is OK, but can you add a comment about this? It's a little non-obvious. You could potentially also ASSERT(!settings_->disabled())

**Comment by htuch** [comment]
@euroelessar we suspect this may have introduced a regression in the execution time for `buffer_filter_integration_test`; could you please take a look? CC @jmarantz

**Comment by jmarantz** [comment]
Specifically, Protocols/BufferIntegrationTest.RouterRequestPopulateContentLength/IPv4_HttpDownstream_HttpUpstream takes 15+ seconds and most of it I think is in `response->waitForEndStream()`

Running with `--gtest_filter=\*BufferIntegrationTest.RouterRequestPopulateContentLength\*` executes 8 variants of the same test and consequently takes 120 seconds.

**Comment by mattklein123** [comment]
Possibly related to https://github.com/envoyproxy/envoy/issues/7872

**Comment by jmarantz** [comment]
Appears to be fixed by https://github.com/envoyproxy/envoy/pull/7885

**Pull request merged**